dffml/operation/deploy/deploy/operations.py

```diff
@@ -0,0 +1,104 @@
+import os
+import json
+import asyncio
+import pathlib
+from typing import List,Dict,Any
+
+from dffml.df.base import op
+from .definitions import *
+from .exceptions import CannotRemoveContainer
+
+from dffml_feature_git.feature.operations import clone_git_repo
+from dffml_feature_git.util.proc import check_output, create, stop
+
+
+@op(
+    inputs={"payload":git_payload},
+    outputs={"url":clone_git_repo.op.inputs["URL"]}
+)
+def get_url_from_payload(payload:Dict[str,Any]):
+    print(f" \n\n get_url_from_payload {payload} ")
+    return {
+        "url": payload["repository"]["clone_url"]
+    }
+
+@op(inputs={"payload":git_payload},outputs={"is_default_branch":is_default_branch})
+def check_if_default_branch(payload):
+    pushed_branch = payload["ref"].split("/")[-1]
+    default_branch = payload["repository"]["default_branch"]
+    print(f" is_default_branch:{default_branch==pushed_branch} ")
+    return {
+        "is_default_branch":(default_branch==pushed_branch)
+    }
+
+@op(inputs={"payload":git_payload},outputs={"image_tag":docker_image_tag})
+def get_image_tag(payload):
+    url = payload["repository"]["html_url"] # eg:-"https://github.com/username/Hello-World",
+    tag = url.split("/")
+    tag = f"{tag[-2]}/{tag[-1]}"
+    print(f"image_tag:{tag} ")
+    return {"image_tag":tag}
+
+@op(
+    inputs={"tag":docker_image_tag},
+    outputs={"running_containers":docker_running_containers}
+)
+async def get_running_containers(tag):
+    containers = await check_output("docker","ps","--filter",f"ancestor={tag}", "--format","{{.ID}}" )
+    containers =[container for container in containers.strip().split("\n") if container ]
+    print(f"running containers : {containers}\n")
+    return {
+        "running_containers":containers
+            }
+
+@op(inputs={"containers":docker_running_containers},outputs={"status":got_running_containers})
+def get_status_running_containes(containers):
```

Comment:
@pdxjohnny reason for `get_status_running_containers`
without it :
![Untitled Diagram](https://user-images.githubusercontent.com/39557075/79630138-72c22a80-816c-11ea-983a-447761b37c56.png)
Now if `docker_build_image` gets run before `get_running_containers` ,it'll tag the newly built image with `org/repo`,and any containers which were already running with an image of this tag gets renamed,causing `get_running_containers` to return an empty list.
with `get_running_status`:
![Untitled Diagram(1)](https://user-images.githubusercontent.com/39557075/79630436-f11fcc00-816e-11ea-8f75-7a932f61c989.png)
We could have the same effect by,having `containers` as an input to `docker_build_image`,but i wanted to make it explicit.

Reply:
> Now if docker_build_image gets run before get_running_containers ,it'll tag the newly built image with org/repo,and any containers which were already running with an image of this tag gets renamed,causing get_running_containers to return an empty list.

Ohhhhhhhh, for some reason I wasn't getting what you were saying yesterday. That makes sense. Let's leave it like you have it, explicit is always best.